scripts/dts/edtlib.py

```diff
@@ -0,0 +1,1628 @@
+the binding for the device.
+
+Bindings are files that describe device tree nodes. Device tree nodes are
+usually mapped to bindings via their 'compatible = "..."' property, but binding
+data can also come from a 'sub-node:' key in the binding for the parent device
+tree node.
+
+The top-level entry point of the library is the EDT class. EDT.__init__() takes
+a .dts file to parse and the path of a directory containing bindings.
+"""
+
+import os
+import re
+import sys
+
+import yaml
+
+from dtlib import DT, DTError, to_num, to_nums
+
+# NOTE: testedtlib.py is the test suite for this library. It can be run
+# directly.
+
+# Implementation notes
+# --------------------
+#
+# A '_' prefix on an identifier in Python is a convention for marking it private.
+# Please do not access private things. Instead, think of what API you need, and
+# add it.
+#
+# This library is layered on top of dtlib, and is not meant to expose it to
+# clients. This keeps the header generation script simple.
+#
+# General biased advice:
+#
+# - Consider using @property for APIs that don't need parameters. It makes
+#   functions look like attributes, which is less awkward in clients, and makes
+#   it easy to switch back and forth between variables and functions.
+#
+# - Think about the data type of the thing you're exposing. Exposing something
+#   as e.g. a list or a dictionary is often nicer and more flexible than adding
+#   a function.
+#
+# - Avoid get_*() prefixes on functions. Name them after the thing they return
+#   instead. This often makes the code read more naturally in callers.
+#
+#   Also, consider using @property instead of get_*().
+#
+# - Don't expose dtlib stuff directly.
+#
+# - Add documentation for any new APIs you add.
+#
+#   The convention here is that docstrings (quoted strings) are used for public
+#   APIs, and "doc comments" for internal functions.
+#
+#   @properties are documented in the class docstring, as if they were
+#   variables. See the existing @properties for a template.
+#
+# - Please use ""-quoted strings instead of ''-quoted strings, just to make
+#   things consistent (''-quoting is more common otherwise in Python)
+
+#
+# Public classes
+#
+
+
+class EDT:
+    """
+    Represents a "high-level" view of a device tree, with a list of devices
+    that each have some number of registers, etc.
+
+    These attributes are available on EDT objects:
+
+    devices:
+      A list of Device objects for the devices
+
+    dts_path:
+      The .dts path passed to __init__()
+
+    bindings_dir:
+      The bindings directory path passed to __init__()
+    """
+    def __init__(self, dts, bindings_dir):
+        """
+        EDT constructor. This is the top-level entry point to the library.
+
+        dts:
+          Path to device tree .dts file
+
+        bindings_dir:
+          Path to directory containing bindings, in YAML format. This directory
+          is recursively searched for .yaml files.
```

Comment:
Should be a list of bindings directories to support applications that have their own bindings (additional to Zephyr).